File: templates/problems/06_linked_list/08_find_duplicate/python/solution.py
def solve_find_duplicate(nums: list[int]) -> int:
    """Find the duplicate number using Floyd's cycle detection.

    Time: O(n), Space: O(1).
    """
    # Phase 1: Find intersection point
    slow = nums[0]
    fast = nums[nums[0]]
    while slow != fast:
        slow = nums[slow]
        fast = nums[nums[fast]]
    # Phase 2: Find entrance to the cycle
    slow = 0
    while slow != fast:
        slow = nums[slow]
        fast = nums[fast]
    return slow

File: templates/problems/06_linked_list/08_find_duplicate/typescript/solution.ts
export function solveFindDuplicate(nums: number[]): number {
    // Phase 1: Find intersection point
    let slow = nums[0];
    let fast = nums[nums[0]];
    while (slow !== fast) {
        slow = nums[slow];
        fast = nums[nums[fast]];
    }
    // Phase 2: Find entrance to the cycle
    slow = 0;
    while (slow !== fast) {
        slow = nums[slow];
        fast = nums[fast];
    }
    return slow;
}
